engine-kommand/src/commonMain/kotlin/com/saveourtool/processbuilder/KommandEngine.kt

```diff
@@ -0,0 +1,66 @@
+package com.saveourtool.processbuilder
+
+import com.saveourtool.processbuilder.utils.CurrentOs
+import com.saveourtool.processbuilder.utils.cwd
+import com.saveourtool.processbuilder.utils.getCurrentOs
+import com.saveourtool.processbuilder.utils.logger
+import com.saveourtool.processbuilder.utils.stderr
+import com.saveourtool.processbuilder.utils.stdout
+
+import com.kgit2.process.Command
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+
+private typealias CommandWithArgs = Pair<String, Array<String>>
+
+object KommandEngine : Engine {
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    override suspend fun execute(
+        command: String,
+        config: ProcessBuilderConfig,
+    ): ExecutionResult {
+        val (cmd, args) = getCommandWithArgs(escapeDoubleTicks(command)).also {
+            logger.info { "${it.first} ${it.second.joinToString(" ")}" }
+        }
+        val stdout = config.stdout
+        val stderr = config.stderr
+        val child = Command(cmd)
+            .args(*args)
+            .cwd(config.workingDirectory)
+            .stdout(stdout)
+            .stderr(stderr)
+            .spawn()
+        return scope.async {
+            val stdoutLines = child.getChildStdout()?.lines()?.toList()
+                .orEmpty()
+                .let(stdout::printToFileIfNeeded)
+            val stderrLines = child.getChildStderr()?.lines()?.toList()
+                .orEmpty()
+                .let(stderr::printToFileIfNeeded)
+            val childExitStatus = child.wait()
+            val executionResult = ExecutionResult(
+                childExitStatus.code,
+                stdoutLines,
+                stderrLines,
+            ).also { result ->
+                if (result.stderr.isNotEmpty()) {
+                    logger.warn { "stderr of `$command`:\t${result.stderr.joinToString("\t")}" }
+                }
+            }
+            executionResult
+        }.await()
+    }
+
+    private fun getCommandWithArgs(command: String): CommandWithArgs = when (getCurrentOs()) {
+        CurrentOs.WINDOWS -> "CMD.EXE" to arrayOf("/C", command)
+        else -> "/bin/sh" to arrayOf("-c", command)
```

Comment:
If you have difficulties quoting your command, you can create a shell script (a temporary file) and run (note the redirection):

```bash
/bin/sh </tmp/your-temporary-script.sh
```

This would allow you to pass the user-supplied command **unmodified**, by just writing it to the script.
Or you can pass the command(s) directly to the `stdin` of the shell using the `Process`/`ProcessBuilder` API &mdash; even w/o the temporary script. I believe this approach is somewhat safer.

For Windows and `cmd.exe`, you'll need a temporary batch file.

Reply:
Discussed offline, decided to leave as is (at list for Linux)